2015/presentations/Business logic with PostgreSQL and Python/text.md

```diff
@@ -0,0 +1,399 @@
+
+I have to make some other clarification here. Term "procedural" and "function" can be a little bit confusing here. Procedural programming is not the only pattern that you can use in plPython. Once you create plpy function then the actual body of it can be pretty complex object. For sure I am not trying to convince you here that object programming in such a case like business logic in DB is the right way. Zen of Python described it as below. Read it again, again, and... again before you start writing actually any Python code.
+
+> Beautiful is better than ugly.
+
+> Explicit is better than implicit.
+
+> Simple is better than complex.
+
+> Flat is better than nested.
+
+> Readability counts.
+
+> If the implementation is hard to explain, it’s a bad idea
+
+What I am trying to say here is that just to be careful with creating plpy functions. Not to heavy, not too complex. Simplicity can bring you a lot of benefits later when you have to squash a bug.
+
+# Hello world
+
+Very basic hello world function in plPython, ex.
+
+    create or replace function logic.hello_world()
+    returns void as
+    $$
+    """
+    Code code code code
+    """
+    plpy.info("hello world")
+    $$
+    LANGUAGE plpythonu VOLATILE;
+
+As you can see from above example you are able to print out some messages. They go to stdout of PostgreSQL directly which means in production stout is redirected to a log file. Depending on PostgreSQL log level you can hide some unwanted messages from plPy function same as some context like which function calls which. Example of a verbose log level.
+
+    pie=# select * from logic.view_and_set_discounted_sales(20, 23) ;
+    INFO:  [view_and_set_discounted_sales] Updateing item id: 36 with new discounted price: 0.529331946972
+    CONTEXT:  PL/Python function "view_and_set_discounted_sales"
+    INFO:  [view_and_set_discounted_sales] Updateing item id: 40 with new discounted price: 5.03912459204
+    CONTEXT:  PL/Python function "view_and_set_discounted_sales"
+    INFO:  [view_and_set_discounted_sales] Updateing item id: 45 with new discounted price: 3.37963644177
+    CONTEXT:  PL/Python function "view_and_set_discounted_sales"
+
+PostgreSQL with above log level (notice) will not only show you your *info* messages but also **context**. That is something that is very helpful with debugging production cases. Once function is being called from context you can see what kind of triggers or sub-functions were called.
+
+As I mentioned before you have an access to all python modules and standard libraries. Below you can see how to access Redis and store some data in it. 
+
+    create or replace function logic.get_active_bills()
+    returns text as
+    $$
+    
+    import redis

```

Comment:
I don't know what kind of standard python you have but mine does not have redis in it? If you're showing how to use redis, show to how to install that library inside PostgreSQL.